src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs

```diff
@@ -0,0 +1,226 @@
+            }
+        }
+
+        protected override Task<bool> IsApplicableImplAsync(IUpgradeContext context, CancellationToken token) => Task.FromResult(context?.CurrentProject is not null);
+
+        protected override Task<UpgradeStepInitializeResult> InitializeImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            try
+            {
+                var currentProject = context.CurrentProject.Required();
+                return Initialize(currentProject);
+            }
+            catch (Exception e)
+            {
+                Logger.LogError(e, "Unexpected error happened during CoreWCF porting.");
+                return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Failed, "Project failed to port to CoreWCF. Please review error message and log file for more information.", BuildBreakRisk.Medium));
+            }
+        }
+
+        protected override Task<UpgradeStepApplyResult> ApplyImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            try
+            {
+                return Apply();
+            }
+            catch (Exception e)
+            {
+                Logger.LogError(e, "Unexpected error happened during CoreWCF porting.");
+                return Task.FromResult(new UpgradeStepApplyResult(UpgradeStepStatus.Failed, "Project failed to port to CoreWCF. Please review error message and log file for more information."));
+            }
+        }
+
+        public Task<UpgradeStepInitializeResult> Initialize(IProject project)
+        {
+            _path = new Dictionary<string, IEnumerable<string>>();
+            FindPath(project);
+            if (_path.Count != 3 && _path.Count != 4)
+            {
+                return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, "Cannot find required file(s) for WCF update. The project is not applicable for automated update and this step is complete.", BuildBreakRisk.Low));
+            }
+
+            // if the project is not applicable, skip the update step
+            if (!WCFUpdateChecker.IsWCFUpdateApplicable(_path, Logger))
+            {
+                var message = "This project is not applicable for updating to CoreWCF since references to WCF were not found. No more work needs to be done and this step is complete.";
+                Logger.LogInformation(message);
+                return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, message, BuildBreakRisk.Low));
+            }
+
+            // construct updaters
+            _configUpdater = UpdaterFactory.GetConfigUpdater(_path["config"].First(), Logger);
```

Comment:
Technically, ILogger's are usually specific to a type (so that it's possible to correlate traces to the types that emitted them) so you shouldn't share this class's ILogger. Instead, you can get an ILoggerFactory from DI and use that to create loggers for other classes, as neede.